### PR TITLE
CRM: Resolves 2930, 2088, 2932 - various fixes for sales funnel

### PR DESCRIPTION
**Pull request by tbradsha**
Resolves Automattic/zero-bs-crm#2930, Automattic/zero-bs-crm#2088, Automattic/zero-bs-crm#2932 - various fixes for sales funnel

## Proposed changes:

TL;DR the funnel library we use is defunct and has plenty of edge-case bugs. I looked into a few options, most of which were outdated or incompatible with our included outdated libraries. As such, I decided to build our own lightweight funnel.

Changes/fixes include:
* The funnel will show properly (based on init settings) if the funnel status setting is blank
* Having 0 contacts in later statuses no longer breaks things.
* Having the same number of contacts in later statuses no longer breaks things
* Sometimes status counts didn't show
* The legend was misplaced
* Screen resizing is a bit more robust
* Lots of code cleanup
* We now use `Lead,Customer` as the default statuses used if settings were blank instead of `Lead,Contacted,Customer,Upsell`

Additions:
* The funnel will show even when there are no contacts
* Hovering over a funnel section will show which status it belongs to
* There's now an `alt` attribute on each funnel section
* Funnel sections support links (not currently used)
* The funnel logic should easily allow for multiple funnels on the same page, as well as easy adaptation when moving to React

For consideration:
* Should we swap out for JP colors? If we do, this might clash with other charts.
* Is there anything that needs escaping? I think we're safe, as it's going through `wp_json_encode()`.
* I've disabled the legend for now with a simple constant, but can add it back if we think that's better.

One can build a funnel like so:

```
// build sales funnel
element = document.getElementById('jpcrm_sales_funnel');
funnel_data = [{"count":"1828","backfill_count":1968,"contact_status":"Lead","link":false},{"count":"131","backfill_count":140,"contact_status":"Customer","link":false},{"count":"4","backfill_count":9,"contact_status":"Blacklisted","link":false},{"count":"5","backfill_count":5,"contact_status":"Refused","link":false},{"count":"0","backfill_count":0,"contact_status":"a","link":false},{"count":"0","backfill_count":0,"contact_status":"b","link":false},{"count":"0","backfill_count":0,"contact_status":"c","link":false},{"count":"0","backfill_count":0,"contact_status":"d","link":false}];
jpcrm_build_funnel(funnel_data,element);
```

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Testing instructions:

Play with various circumstances, including:

* 0 contacts in status 1, 2, and 3, but 2 contacts in status 4
* 0 funnel statuses
* 1 funnel status
* LOTS of funnel statuses
* Window resizing
* 1 contact in status 1 and 0 contacts in status 2
* Hiding the panel (dashboard setting in upper right)

Ensure that nothing is worse than it was, and that most things are better.

### CODE REVIEW

File: projects/plugins/crm/js/jpcrm-admin-funnel.js [hunk 1]
```diff
@@ -0,0 +1,57 @@
+/**
+ * Draws a simple funnel
+ * 
+ * Inspired in part by https://codepen.io/tylerlwsmith/pen/RwNZKgp
+ * 
+ * @param funnel_data array of funnel data (count, backfill_count, contact_status, and link)
```
Comment (gogdzl) [inline projects/plugins/crm/js/jpcrm-admin-funnel.js:34]: There is no param named funnel_data

File: projects/plugins/crm/js/jpcrm-admin-funnel.js [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+ * @element HTML element to turn into a funnel
+ **/
+
+function jpcrm_build_funnel(data, funnel_element) {
+	// something's wrong, so stop
+	if (!element || !data) return;
```
Comment (gogdzl) [inline projects/plugins/crm/js/jpcrm-admin-funnel.js]: Is `element` meant to be `funnel_element`  ?

File: projects/plugins/crm/js/jpcrm-admin-funnel.js [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+	const COLORS = [ '#000', '#222', '#333', '#035d88', '#0073aa', '#00a0d2' ];
+
+	// show legend if desired
+	const SHOW_FUNNEL_LEGEND = false;
+
+	num_of_funnel_sections = data.length;
+	funnel_height = num_of_funnel_sections * 50;
```
Comment (gogdzl) [inline projects/plugins/crm/js/jpcrm-admin-funnel.js]: I think that not using `let` here declares these two variables in the global scope. I'm not sure if this is a problem or not, I just want to highlight this.

File: projects/plugins/crm/js/jpcrm-admin-funnel.js (hunk 2 above)
Comment (gogdzl) [reply projects/plugins/crm/js/jpcrm-admin-funnel.js]: Just read again the file, `element` is being declared in the global scope, and at the same time being passed as the function's param named `funnel_element`.

**Review by gogdzl — CHANGES_REQUESTED** [review]
I like where this is going!
There are some comments about the new JS function, it seems that there is some slight confusion with the params and variables that could be improved.

Also, while spinning up a new Jurassic Ninja, the default colors are black for both Lead, and Customer status. Could we default it to those shades of blue?

image

File: projects/plugins/crm/js/jpcrm-admin-funnel.js (hunk 1 above)
Comment (tbradsha) [reply projects/plugins/crm/js/jpcrm-admin-funnel.js:34]: Oops! 7f06270775

File: projects/plugins/crm/js/jpcrm-admin-funnel.js (hunk 2 above)
Comment (tbradsha) [reply projects/plugins/crm/js/jpcrm-admin-funnel.js]: Indeed; the initial prototype wasn't in a function, and I forgot to fix this. 😅 b9fea3235d

File: projects/plugins/crm/js/jpcrm-admin-funnel.js (hunk 3 above)
Comment (tbradsha) [reply projects/plugins/crm/js/jpcrm-admin-funnel.js]: Indeed, better to limit scope: 163f7124f2

**Comment by tbradsha** [comment]
>There are some comments about the new JS function, it seems that there is some slight confusion with the params and variables that could be improved.

All valid comments, and I believe I've addressed them all now.

> Also, while spinning up a new Jurassic Ninja, the default colors are black for both Lead, and Customer status. Could we default it to those shades of blue?

Technically it's black and dark grey (`#000` vs. `#222`). These were the colors used previously, though it looks like they were in reverse order before, and most people didn't have that many statuses in the filter so wouldn't see the darker ones.  I can reverse it so it matches the old colors, or we can choose a new set (e.g. https://color-studio.blog/ ). Thoughts?

**Comment by tbradsha** [comment]
I've adjusted the funnel to go from light to dark:

image

There are 10 predefined colors, pulled from here: https://color-studio.blog/

As such:
* If there are 10 funnel segments, it returns them all.
* If there are less than 10 funnel segments, it returns `n` colors from the midpoint of the colors array so we don't weight too light or two dark.
* If there are more than 10 funnel segments, it returns the full array repeated.

**Comment by gogdzl** [comment]
Our first sight of the funnel looks really good now:

image


Showing all default statuses in the same fresh install:

image

With some data:

image

**Review by gogdzl — DISMISSED** [review]
Superb job! LGTM!

File: projects/plugins/crm/js/jpcrm-admin-funnel.js [hunk 4]
```diff
@@ -6,23 +6,24 @@
  * @returns an array of colors
  **/
 function jpcrm_get_segment_colors(num_segments) {
-	let colors = [ '#68b3e8', '#399ce3', '#1689db', '#0675c4', '#055d9c', '#044b7a', '#02395c', '#01283d', '#001621', '#000000' ];
+	const COLORS = [ '#68b3e8', '#399ce3', '#1689db', '#0675c4', '#055d9c', '#044b7a', '#02395c', '#01283d', '#001621', '#000000' ];
 
-	if (num_segments === colors.length) {
+	if (num_segments === COLORS.length) {
 		// use full set of colors
 		return colors;
```
Comment (gogdzl) [inline projects/plugins/crm/js/jpcrm-admin-funnel.js]: I think here we are supposed to return `COLORS`.

Comment (tbradsha) [reply projects/plugins/crm/js/jpcrm-admin-funnel.js]: Sigh... 87ed55a639

**Review by gogdzl — APPROVED** [review]
image

**Pull request merged**